snapcraft/internal/project_loader.py

```diff
@@ -351,3 +390,22 @@ def _ensure_grade_default(yaml_data, schema):
         logger.warning('"grade" property not specified: defaulting '
                        'to "stable"')
         yaml_data['grade'] = schema['grade']['default']
+
+
+def _expand_filesets_for(step, properties):
+    filesets = properties.get('filesets', {})
+    fileset_for_step = properties.get(step, {})
+    new_step_set = []
+
+    for item in fileset_for_step:
+        if item.startswith('$'):
+            try:
+                new_step_set.extend(filesets[item[1:]])
+            except KeyError:
+                raise SnapcraftLogicError(
+                    '\'{}\' referred to in the \'{}\' fileset but it is not '
```

Comment:
I am going to ignore this one for now and rework these logic errors as a whole into `snapcraft.internal.errors`